File: book_kit/src/main/java/net/lzbook/kit/utils/CustomWebClient.java
package net.lzbook.kit.utils;

import net.lzbook.kit.app.BaseBookApplication;

import android.annotation.SuppressLint;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.net.http.SslError;
import android.view.View;
import android.webkit.HttpAuthHandler;
import android.webkit.SslErrorHandler;
import android.webkit.WebSettings;
import android.webkit.WebSettings.LayoutAlgorithm;
import android.webkit.WebSettings.RenderPriority;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.androidquery.util.Common;
import com.dingyue.contract.util.CommonUtil;

public class CustomWebClient extends WebViewClient {
    private final static String TAG = "CustomWebClient";
    public WebSettings webSetting;
    public int ERROR_CODE = 0;
    public int pageStartCount = 0;
    Context context;
    WebView webview;
    onErrorCallback onError;
    onStartedCallback onLoad;
    onFinishedCallback onFinished;
    onOverrideCallback override;
    int pageFinishCount = 0;

    public CustomWebClient(Context context, WebView webview) {
        super();
        this.context = context;
        this.webview = webview;
    }

    /**
     * return true , application handle url itself, do not use current WebView
     * return false,current WebView handle url, defult: false
     */

    @Override
    public boolean shouldOverrideUrlLoading(final WebView view, String url) {
        // 如下方案可在非微信内部WebView的H5页面中调出微信支付
        if(url.startsWith("weixin://wap/pay?")) {
            try {
                Intent intent = new Intent();
                intent.setAction(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                context.startActivity(intent);
                return true;
            } catch (Exception exception) {
                if (exception instanceof ActivityNotFoundException) {
                    CommonUtil.showToastMessage("未找到微信客户端，请先安装微信！");
                }
                exception.printStackTrace();
                return true;
            }
        }

        if (url.startsWith("alipays://platformapi/startApp?")) {
            try {
                Intent intent = new Intent();
                intent.setAction(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                context.startActivity(intent);
                return true;
            } catch (Exception exception) {
                if (exception instanceof ActivityNotFoundException) {
                    CommonUtil.showToastMessage("未找到支付宝客户端，请先安装支付宝！");
                }
                exception.printStackTrace();
                return  super.shouldOverrideUrlLoading(view, url);
            }
        }
        return super.shouldOverrideUrlLoading(view, url);
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        pageStartCount++;
        AppLog.d(TAG, "pageStartCount  " + pageStartCount + " ERROR_CODE : " + ERROR_CODE + " onError : " + onError);
        if (onLoad != null && pageStartCount == 1) {
            onLoad.onLoadStarted(url);// FIXME 保证回调只执行一次
        }
        if (webview != null && pageStartCount == 1) {
            webview.setVisibility(View.GONE);
        } else if (ERROR_CODE != 0 && onError != null) {
            if (webview != null) {
                webview.stopLoading();
                webview.clearView();
                webview.setVisibility(View.GONE);// 出错时不显示Webview的错误页面，由app处理
            }
            onError.onErrorReceived();
        }
        AppLog.e(TAG, "onPageStarted view " + " ERROR_CODE : " + ERROR_CODE + " onError : " + onError + view + " webview " + webview);//初始化webview与加载webview要一致
        super.onPageStarted(view, url, favicon);
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        pageFinishCount++;
        AppLog.d(TAG, "onPageFinished ERROR_CODE " + ERROR_CODE + onFinished);
        if (ERROR_CODE == 0 && onFinished != null) {// FIXME
            onFinished.onLoadFinished();
            if (webview != null) {

                webview.setVisibility(View.VISIBLE);
                // FIXME 网页显示之前，空白一段时间
            }
            if (webSetting != null) {
                webSetting.setBlockNetworkImage(false);// 通过图片的延迟载入，让网页能更快地显示
            }
        } else if (ERROR_CODE != 0 && onError != null) {
            if (webview != null) {
                webview.clearView();
                webview.stopLoading();
                webview.setVisibility(View.GONE);// 出错时不显示Webview的错误页面，由app处理
            }
            onError.onErrorReceived();
        }
        AppLog.d(TAG, "onPageFinished view" + view + " webview " + webview + " ERROR_CODE : " + ERROR_CODE + " onError : " + onError);
        super.onPageFinished(view, url);
    }

    @Override
    public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
        ERROR_CODE = errorCode;// FIXME 标识是否出错
        AppLog.e(TAG, "error[" + errorCode + " --- " + failingUrl + "] " + description);
        if (webview != null) {
            webview.clearView();
            webview.stopLoading();
        }
        super.onReceivedError(view, errorCode, description, failingUrl);
    }

    @Override
    public void onReceivedHttpAuthRequest(WebView view, HttpAuthHandler handler, String host, String realm) {
        super.onReceivedHttpAuthRequest(view, handler, host, realm);
    }

    /**
     * 处理https请求
     */
    public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
        handler.proceed();
    }

    // =================================================
    // activity 回调
    // ============================================
    public void doClear() {
        if (ERROR_CODE != 0) {// 解决出现过error情况后重新加载ERROR_CODE 全局变量不变
            ERROR_CODE = 0;
        }
        AppLog.d(TAG, "doClear ERROR_CODE" + ERROR_CODE);
    }

    public void setErrorAction(onErrorCallback onError) {
        this.onError = onError;
    }

    public void setStartedAction(onStartedCallback onStarted) {
        this.onLoad = onStarted;
    }

    public void setFinishedAction(onFinishedCallback onFinished) {
        this.onFinished = onFinished;
    }

    @SuppressLint("SetJavaScriptEnabled")
    public void setWebSettings() {
        // webview.requestFocus();
        String appCachePath = null;
        String dbPath = null;
        if (webview != null && webSetting == null) {
            webview.setScrollBarStyle(WebView.SCROLLBARS_INSIDE_OVERLAY);
            webSetting = webview.getSettings();
        }

        if (context == null) {
            context = BaseBookApplication.getGlobalContext();
        }
        if (context != null && context.getCacheDir() != null && context.getDir("databases", 0) != null) {

            try {
                appCachePath = context.getCacheDir().getAbsolutePath();
                dbPath = context.getDir("databases", 0).getPath();
            } catch (NullPointerException e) {
                e.printStackTrace();
            }

            AppLog.d(TAG, "appCachePath " + appCachePath);
            AppLog.d(TAG, "dbPath " + dbPath);
        }
        if (webSetting != null && appCachePath != null)
            webSetting.setAppCachePath(appCachePath);
        if (webSetting != null && dbPath != null) {
            webSetting.setDatabasePath(dbPath);// TODO
        }
        if (webSetting != null) {

            if (NetWorkUtils.NETWORK_TYPE == NetWorkUtils.NETWORK_NONE) {
                webSetting.setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);
            } else {
                webSetting.setCacheMode(WebSettings.LOAD_DEFAULT);
            }
            // HTML5 API flags
            webSetting.setAppCacheEnabled(true);
            webSetting.setDatabaseEnabled(true);
            webSetting.setDomStorageEnabled(true);

            // HTML5 configuration parametersettings.

            webSetting.setAllowFileAccess(true);
            webSetting.setAppCacheMaxSize(1024 * 1024 * 8);// TODO

            // WebView inside Browser doesn't want initial focus to be set.
            webSetting.setNeedInitialFocus(false);
            // Browser supports multiple windows
            webSetting.setSupportMultipleWindows(true);

            try {
                webSetting.setJavaScriptEnabled(true);
            } catch (Exception e) {
                e.printStackTrace();
            }

            webSetting.setJavaScriptCanOpenWindowsAutomatically(false);
            webSetting.setLoadsImagesAutomatically(true);
            webSetting.setLayoutAlgorithm(LayoutAlgorithm.NORMAL);
            webSetting.setRenderPriority(RenderPriority.HIGH);// 提高渲染的优先级

            webSetting.setBlockNetworkImage(false); // 图片下载阻塞
            webSetting.setUseWideViewPort(true);//FIXME
        }

    }

    public interface onErrorCallback {
        public void onErrorReceived();
    }

    public interface onStartedCallback {
        public void onLoadStarted(String url);
    }

    public interface onFinishedCallback {
        public void onLoadFinished();
    }

    public interface onOverrideCallback {
        public boolean onOverride(WebView view, String url);
    }
}

File: book_kit/src/main/java/net/lzbook/kit/utils/web/CustomWebClient.kt
package net.lzbook.kit.utils.web

import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.net.http.SslError
import android.os.Build
import android.support.annotation.RequiresApi
import android.view.View
import android.webkit.*
import android.webkit.WebSettings.LayoutAlgorithm
import android.webkit.WebSettings.RenderPriority
import com.ding.basic.util.sp.SPKey
import com.ding.basic.util.sp.SPUtils
import com.orhanobut.logger.Logger
import net.lzbook.kit.app.base.BaseBookApplication
import net.lzbook.kit.utils.NetWorkUtils
import net.lzbook.kit.utils.toast.ToastUtil

class CustomWebClient(var context: Context?, internal var webView: WebView?) : WebViewClient() {

    private var loadingStartCount = 0
    private var loadingFinishCount = 0

    private var loadingErrorCount = 0

    private var webSettings: WebSettings? = null

    private var webResourceCache = WebResourceCache.loadWebResourceCache()

    /***
     * WebView加载开始监听
     * **/
    private var loadingWebViewStart: ((url: String?) -> Unit)? = null

    /***
     * WebView加载完成监听
     * **/
    private var loadingWebViewFinish: (() -> Unit)? = null

    /***
     * 加载WebView异常监听
     * **/
    private var loadingWebViewError: (() -> Unit)? = null

    /***
     * WebView每次加载开始监听
     * **/
    private var loadingEveryWebViewStart: ((url: String?) -> Unit)? = null

    init {
        this.webView?.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, newProgress: Int) {
                super.onProgressChanged(view, newProgress)
                loadingWebViewFinish()
            }
        }
    }

    companion object {
        private val staticResourceRule = SPUtils.getOnlineConfigSharedString(SPKey.DY_STATIC_RESOURCE_RULE)
        val interceptHostList = staticResourceRule.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    }

    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        if (url.startsWith("weixin://wap/pay?")) {
            return try {
                val intent = Intent()
                intent.action = Intent.ACTION_VIEW
                intent.data = Uri.parse(url)
                context?.startActivity(intent)
                true
            } catch (exception: Exception) {
                if (exception is ActivityNotFoundException) {
                    ToastUtil.showToastMessage("未找到微信客户端，请先安装微信！")
                }
                exception.printStackTrace()
                true
            }
        } else if (url.startsWith("alipays://platformapi/startApp?")) {
            return try {
                val intent = Intent()
                intent.action = Intent.ACTION_VIEW
                intent.data = Uri.parse(url)
                context?.startActivity(intent)
                true
            } catch (exception: Exception) {
                if (exception is ActivityNotFoundException) {
                    ToastUtil.showToastMessage("未找到支付宝客户端，请先安装支付宝！")
                }
                exception.printStackTrace()
                true
            }
        }
        return super.shouldOverrideUrlLoading(view, url)
    }

    override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
        loadingStartCount++

        if (loadingWebViewStart != null && loadingStartCount == 1) {
            loadingWebViewStart?.invoke(url)
        }

        if (webView != null && loadingStartCount == 1) {
            webView?.visibility = View.GONE
        } else if (loadingErrorCount != 0 && loadingWebViewError != null) {
            loadingWebViewError()
        }
        if(loadingEveryWebViewStart != null){
            loadingEveryWebViewStart?.invoke(url)
        }
        super.onPageStarted(view, url, favicon)
    }

    override fun onPageFinished(view: WebView?, url: String?) {
        loadingFinishCount++
        if (loadingErrorCount == 0 && loadingWebViewFinish != null) {
           loadingWebViewFinish()
        } else if (loadingErrorCount != 0 && loadingWebViewError != null) {
            loadingWebViewError()
        }
        super.onPageFinished(view, url)
    }

    override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
        loadingErrorCount = errorCode
        loadingWebViewError()
        super.onReceivedError(view, errorCode, description, failingUrl)
    }

    override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
        handler.proceed()
    }

//    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
//    override fun shouldInterceptRequest(view: WebView?, request: WebResourceRequest?): WebResourceResponse? {
//        return interceptWebViewRequest(view, request)
//    }

    /***
     * WebView加载进度
     * **/
    private fun loadingWebViewFinish() {
        if (null != webView) {
            if (webView?.progress ?: 0 > 60) {
                loadingWebViewFinish?.invoke()
            }

            webView?.visibility = View.VISIBLE
        }

        webSettings?.blockNetworkImage = false
    }

    /***
     * WebView加载失败
     * **/
    private fun loadingWebViewError() {
        if (webView != null) {
            webView?.clearView()
            webView?.stopLoading()
            webView?.visibility = View.GONE
        }
        loadingWebViewError?.invoke()
    }


    /***
     * 设置加载WebView开始监听
     * **/
    fun setLoadingWebViewStart(loadingWebViewStart: (url: String?) -> Unit) {
        this.loadingWebViewStart = loadingWebViewStart
    }

    /***
     * 设置每次加载WebView开始监听
     * **/
    fun setLoadingEveryWebViewStart(loadingEveryWebViewStart: (url: String?) -> Unit) {
        this.loadingEveryWebViewStart = loadingEveryWebViewStart
    }

    /***
     * 设置WebView加载完成监听
     * **/
    fun setLoadingWebViewFinish(loadingWebViewFinish: () -> Unit) {
        this.loadingWebViewFinish = loadingWebViewFinish
    }

    /***
     * 设置加载WebView异常监听
     * **/
    fun setLoadingWebViewError(loadingWebViewError: () -> Unit) {
        this.loadingWebViewError = loadingWebViewError
    }

    /***
     * 拦截WebView请求
     * **/
    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun interceptWebViewRequest(view: WebView?, request: WebResourceRequest?): WebResourceResponse? {
        if (request != null && request.url != null && request.url.toString().isNotEmpty()) {
            val url = request.url.toString()
            val webResourceCached = webResourceCache.loadWebViewCache(url)

            if (webResourceCached != null) {
                Logger.e("命中缓存: $url")
                return WebResourceResponse(webResourceCached.mimeType, webResourceCached.encoding, webResourceCached.file.inputStream())
            }

            return when {
                url.startsWith("http") -> {
                    val host = request.url.host
                    when {
                        interceptHostList.contains(host) -> handleInterceptRequest(view, request, url)
                        else -> super.shouldInterceptRequest(view, request)
                    }
                }
                url.startsWith("file://") -> handleInterceptRequest(view, request, url)
                else -> return super.shouldInterceptRequest(view, request)
            }
        } else {
            return super.shouldInterceptRequest(view, request)
        }
    }

    /***
     * 处理拦截请求
     * **/
    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun handleInterceptRequest(view: WebView?, request: WebResourceRequest?, url: String): WebResourceResponse? {
        val fileExtension = MimeTypeMap.getFileExtensionFromUrl(url)
        val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension)

        return if (fileExtension.isNotEmpty()) {
            if ((mimeType == "image/jpeg" || mimeType == "image/png")) {
                webResourceCache.handleImageRequest(url, mimeType) ?: super.shouldInterceptRequest(view, request)
            } else if (fileExtension == "js" || fileExtension == "css") {
                return webResourceCache.handleOtherRequest(url, mimeType) ?: super.shouldInterceptRequest(view, request)
            } else {
                super.shouldInterceptRequest(view, request)
            }
        } else {
            super.shouldInterceptRequest(view, request)
        }
    }

    /***
     * 初始化WebView设置
     * **/
    @SuppressLint("SetJavaScriptEnabled")
    fun initWebViewSetting() {
        var cachePath: String? = null
        var databasePath: String? = null

        if (webView != null && webSettings == null) {
            webView?.scrollBarStyle = WebView.SCROLLBARS_INSIDE_OVERLAY
            webSettings = webView?.settings
        }

        if (context == null) {
            context = BaseBookApplication.getGlobalContext()
        }

        if (context != null && context?.cacheDir != null && context!!.getDir("databases", 0) != null) {
            try {
                cachePath = context?.cacheDir?.absolutePath
                databasePath = context?.getDir("databases", Context.MODE_PRIVATE)?.path
            } catch (exception: Exception) {
                exception.printStackTrace()
            }
        }

        if (webSettings != null && cachePath != null && cachePath.isNotEmpty()) {
            webSettings?.setAppCachePath(cachePath)
        }

        if (webSettings != null && databasePath != null && databasePath.isNotEmpty()) {
            webSettings?.databasePath = databasePath
        }

        if (NetWorkUtils.NETWORK_TYPE == NetWorkUtils.NETWORK_NONE) {
            //如果有本地缓存则直接使用本地缓存，而不管缓存数据是否过期失效，否则加载网络数据
            webSettings?.cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK
        } else {
            //如果有本地缓存，且缓存有效未过期，则直接使用本地缓存，否则加载网络数据
            webSettings?.cacheMode = WebSettings.LOAD_DEFAULT
        }

        webSettings?.setAppCacheEnabled(true)

        webSettings?.databaseEnabled = true

        webSettings?.domStorageEnabled = true

        webSettings?.allowFileAccess = true

        webSettings?.setAppCacheMaxSize((1024 * 1024 * 15).toLong())

        webSettings?.setNeedInitialFocus(false)

        webSettings?.setSupportMultipleWindows(true)

        try {
            webSettings?.javaScriptEnabled = true
        } catch (exception: Exception) {
            exception.printStackTrace()
        }

        webSettings?.javaScriptCanOpenWindowsAutomatically = false

        webSettings?.loadsImagesAutomatically = true

        webSettings?.layoutAlgorithm = LayoutAlgorithm.NORMAL

        webSettings?.setRenderPriority(RenderPriority.HIGH)

        webSettings?.blockNetworkImage = false

        webSettings?.useWideViewPort = true

        webSettings?.textZoom = 100
    }

    /***
     * 初始化部分数据
     * **/
    fun initParameter() {
        if (loadingErrorCount != 0) {
            loadingErrorCount = 0
        }
    }
}